### PR TITLE
Add advertisement modules and core scaffolding

### PR DESCRIPTION
**Pull request by ellenhp**
Big PR, but this brings 3.5.1-ramatak up to date with my latest work.

Building:

- [ ] Test this script prior to marking as ready for review.

```
# May need adjustment for your specific situation/preferences
export VERSION=3.5.1.stable
mkdir gdworkspace
cd gdworkspace
export GDWORKSPACE=$PWD
# Reproducibility note: HEAD as of opening of this PR is c9b31bdb8fa5727e023eb9f4a7fc4185f541fcf2
git clone https://github.com/ramatakinc/ramatak-proprietary-modules
# Reproducibility note: HEAD as of opening of this PR is 853f15ec3e5104dc2a89e868e3e490b88509a6f9
git clone https://github.com/ramatakinc/android-ad-plugins
cd android-ad-plugins
./gradlew assemble
cd ../godot
scons custom_modules=../ramatak-proprietary-modules debug_symbols=yes
scons custom_modules=../ramatak-proprietary-modules platform=android arch=arm64 target=release_debug
pushd platform/android/java
./gradlew generateGodotTemplates
popd
mkdir -p bin/templates/$VERSION/
ln -s $GDWORKSPACE/android-ad-plugins/admob/build/outputs/aar/admob-debug.aar bin/admob.aar
ln -s $GDWORKSPACE/android-ad-plugins/meta/build/outputs/aar/meta-debug.aar bin/meta.aar
ln -s $PWD/bin/android_source.zip bin/templates/$VERSION/android_source.zip
```

Test Plan:

- [ ] Create new project
- [ ] Export project for android and confirm no errors during build or startup.
- [ ] Enter admob application ID and enable admob in the priorities screen. Export again and confirm no errors during build or startup (there will be a crash on startup if you forget to populate the Application ID).
- [ ] Add 3 ad units (interstitial, rewarded and banner) and populate each with the plugin-specific IDs in the ad units settings editor.

Confirm the following:

- [ ] Admob Application ID and ad unit settings for both plugins are persisted across editor restarts.
- [ ] BannerAdController only lists the banner ad unit as an available option in the inspector.
- [ ] AdController only lists the interstitial and rewarded ads as available options in the inspector.

For the following checks, switch between plugins by adjusting the priorities in Project Settings -> Ramatak -> Priorities.

Add a banner ad controller and toggle auto-show on, export project and verify the banner ad loads on startup (may take a few seconds). Confirm on:

- [ ] Meta
- [ ] Admob

Create two buttons to show interstitial and rewarded ads, respectively. Confirm for all three ad types and both plugins that only test ads are shown for debug builds.

- [ ] Meta Banner
- [ ] Admob Banner
- [ ] Meta Interstitial
- [ ] Admob Interstitial
- [ ] Meta Rewarded
- [ ] Admob Rewarded

Connect all signals for all ad types and confirmed they are fired correctly for all ad types and plugins. Clicked ads will only fire a signal once the user returns to the app, for complicated reasons. We'll deal with that when we create an analytics system.

- [ ] Meta Banner
- [ ] Admob Banner
- [ ] Meta Interstitial
- [ ] Admob Interstitial
- [ ] Meta Rewarded
- [ ] Admob Rewarded

TBD: Confirm that prod ads are used in release builds, if Admob even allows that? I don't want my google account deactivated and they're very serious about ad fraud as I understand it.

### CODE REVIEW

File: core/os/dir_access.cpp [hunk 1]
```diff
@@ -327,13 +327,20 @@ Error DirAccess::copy(String p_from, String p_to, int p_chmod_flags) {
 		size -= bytes_read;
 	}
 
+	fdst->close();
```
Comment (ellenhp) [inline core/os/dir_access.cpp]: These changes were necessary to get the dir copy working which I needed to keep the `user://` build template immutable. I made these changes before learning that in 3.x `user://` is still specific to each project. I think this behavior is what users would expect, but it still changes the API so I want to get rid of it before release. I propose one of two things, depending on 4.x behavior:
* If 4.x behavior is identical (`user://` is specific to a project), we remove these changes, remove the copy, simplify export_plugin.cpp and use the `user://` build template mutably during export.
* If 4.x behavior is different, and `user://` is shared between all projects, we move these changes to another method within `DirAccess`. Whether that method should be exposed to script is an open question.

File: platform/android/export/export_plugin.cpp [hunk 2]
```diff
@@ -2272,58 +2256,15 @@ bool EditorExportPlatformAndroid::can_export(const Ref<EditorExportPreset> &p_pr
...
-	if (int(p_preset->get("custom_build/export_format")) == EXPORT_FORMAT_AAB &&
-			!custom_build_enabled) {
-		valid = false;
-		err += TTR("\"Export AAB\" is only valid when \"Use Custom Build\" is enabled.");
-		err += "\n";
-	}
-
 	// Check the min sdk version.
 	String min_sdk_str = p_preset->get("custom_build/min_sdk");
+	// TODO: Actually parse this and upstream the fix.
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp]: I actually misunderstood the original logic (it's correct), so I should revert these changes before release.

File: platform/android/export/export_plugin.cpp [hunk 3]
```diff
@@ -2342,11 +2283,11 @@ List<String> EditorExportPlatformAndroid::get_binary_extensions(const Ref<Editor
 	return list;
 }
 
-void EditorExportPlatformAndroid::_update_custom_build_project() {
+void EditorExportPlatformAndroid::_update_custom_build_project(const String &p_template_path) {
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp:2323]: These changes can be simplified if we choose to use the `user://` build template mutably.

File: platform/android/export/export_plugin.cpp [hunk 4]
```diff
@@ -2937,509 +2880,219 @@ Error EditorExportPlatformAndroid::export_project_helper(const Ref<EditorExportP
...
-			copy_command = vformat("copyAndRename%sAab", build_type);
-		} else if (export_format == EXPORT_FORMAT_APK) {
-			copy_command = vformat("copyAndRename%sApk", build_type);
-		}
-
-		copy_args.push_back(copy_command);
-
-		copy_args.push_back("-p"); // argument to specify the start directory.
-		copy_args.push_back(build_path); // start directory.
+		dir_access->copy_dir(globalized_template_path, template_path);
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp]: Note: This is the copy that prevents the build template from being used mutably.

File: servers/ramatak/ad_server.h [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+class AdPlugin;
+
+class AdServer : public Object {
+	GDCLASS(AdServer, Object);
+
+public:
+	// Add new ad types to the end of this list before *_MAX to avoid messing up existing projects.
+	enum AdType {
+		AD_TYPE_BANNER,
+		AD_TYPE_INTERSTITIAL,
```
Comment (ellenhp) [inline servers/ramatak/ad_server.h:19]: TODO: Figure out how to handle interstitial/rewarded video ads.

File: servers/ramatak/ad_server.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+		AD_TYPE_REWARDED,
+		AD_TYPE_MAX,
+	};
+
+	enum BannerAdSize {
+		BANNER_SIZE_SMALL,
```
Comment (ellenhp) [inline servers/ramatak/ad_server.h:25]: TODO: Figure out how to deal with plugin-specific size formats, e.g. [Admob adaptive banners](https://developers.google.com/admob/android/banner/anchored-adaptive).

File: servers/ramatak/ad_server.cpp [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		}
+	}
+	WARN_PRINT_ONCE("No ad plugin is capable of handling ad unit.");
+	// There might be a better return value here.
+	return Variant();
+}
+
+Variant AdServer::show_banner(String p_ad_unit, BannerAdSize p_size, BannerAdLocation p_location) {
+	// This looks gross, and it is, but these values are going to cross an FFI boundary into a plugin and
+	// it's much less error-prone to pass a string than an integer, especially because these plugins might
```
Comment (ellenhp) [inline servers/ramatak/ad_server.cpp]: TODO: clarify that I mean an integer representation of an unstable enum that might change from one minor release to the next.

File: platform/android/export/export_plugin.cpp [hunk 8]
```diff
@@ -2800,30 +2741,30 @@ Error EditorExportPlatformAndroid::sign_apk(const Ref<EditorExportPreset> &p_pre
 	return OK;
 }
 
-void EditorExportPlatformAndroid::_clear_assets_directory() {
+void EditorExportPlatformAndroid::_clear_assets_directory(const String &p_template_path) {
 	DirAccessRef da_res = DirAccess::create(DirAccess::ACCESS_RESOURCES);
 
 	// Clear the APK assets directory
-	if (da_res->dir_exists(APK_ASSETS_DIRECTORY)) {
+	if (da_res->dir_exists(p_template_path + APK_ASSETS_DIRECTORY)) {
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp]: Here, above, and below. Replace all concatenations with `plus_file` calls.

File: platform/android/export/export_plugin.cpp [hunk 9]
```diff
@@ -184,8 +185,8 @@ static const char *SPLASH_IMAGE_EXPORT_PATH = "res/drawable-nodpi/splash.png";
 static const char *LEGACY_BUILD_SPLASH_IMAGE_EXPORT_PATH = "res/drawable-nodpi-v4/splash.png";
 static const char *SPLASH_BG_COLOR_PATH = "res/drawable-nodpi/splash_bg_color.png";
 static const char *LEGACY_BUILD_SPLASH_BG_COLOR_PATH = "res/drawable-nodpi-v4/splash_bg_color.png";
-static const char *SPLASH_CONFIG_PATH = "res://android/build/res/drawable/splash_drawable.xml";
-static const char *GDNATIVE_LIBS_PATH = "res://android/build/libs/gdnativelibs.json";
+static const char *SPLASH_CONFIG_PATH = "build/res/drawable/splash_drawable.xml";
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp:188]: TODO: Test that splash screen changes are reflected in exported builds.

File: platform/android/export/export_plugin.cpp [hunk 10]
```diff
@@ -184,8 +185,8 @@ static const char *SPLASH_IMAGE_EXPORT_PATH = "res/drawable-nodpi/splash.png";
...
+static const char *GDNATIVE_LIBS_PATH = "build/libs/gdnativelibs.json";
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp:189]: TODO: Test that gdnative libs are included correctly in exported builds.

File: platform/android/export/export_plugin.cpp [hunk 11]
```diff
@@ -717,7 +720,7 @@ Error EditorExportPlatformAndroid::copy_gradle_so(void *p_userdata, const Shared
 		int abi_index = abis.find(p_so.tags[i]);
 		if (abi_index != -1) {
 			exported = true;
-			String base = "res://android/build/libs";
+			String base = export_data->template_directory + "/build/libs";
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp]: Here, above and below. Replace all concatenations with `plus_file` calls.

File: editor/ramatak/ramatak_settings_editor.cpp [hunk 12]
```diff
@@ -0,0 +1,593 @@
...
+	Dictionary ad_units_config = ProjectSettings::get_singleton()->get("ramatak/monetization/ad_units");
+	Vector<int> selected_items = edit_items_list->get_selected_items();
+	if (selected_items.size() > 0) {
+		// DEV_ASSERT(ad_unit_editor->get_ad_unit() != "");
+		ad_units_config.erase(ad_unit_editor->get_ad_unit());
+		edit_items_list->remove_item(selected_items.get(0));
+		ad_unit_editor->clear();
+		ad_unit_editor->set_disabled(true);
+	}
+	WARN_PRINT("REMOVING");
```
Comment (ellenhp) [inline editor/ramatak/ramatak_settings_editor.cpp]: Remove debug print.

File: editor/ramatak/ramatak_settings_editor.cpp [hunk 13]
```diff
@@ -0,0 +1,593 @@
...
+	Dictionary this_plugin_config = ad_plugin_configs[plugin_id];
+
+	Ref<AdPlugin> plugin = AdServer::get_singleton()->get_plugin_raw(plugin_id);
+	Array key_tooltip_pairs = plugin->get_config_key_tooltip_pairs();
+	for (int i = 0; i < key_tooltip_pairs.size(); i++) {
+		Array pair = key_tooltip_pairs[i];
+		String key = pair[0];
+
+		this_plugin_config[key] = edits[i]->get_text();
+		WARN_PRINT(key);
```
Comment (ellenhp) [inline editor/ramatak/ramatak_settings_editor.cpp]: Here and below. Remove debug print.

File: platform/android/export/export_plugin.cpp [hunk 14]
```diff
@@ -2937,509 +2880,219 @@ Error EditorExportPlatformAndroid::export_project_helper(const Ref<EditorExportP
...
+	} else {
+		DirAccess *dir_access = DirAccess::create(DirAccess::ACCESS_FILESYSTEM);
+		template_path = ProjectSettings::get_singleton()->globalize_path("user://ramatak/tmp/android_build");
+		dir_access->remove_file_or_error(template_path);
+		String globalized_template_path = ProjectSettings::get_singleton()->globalize_path(get_template_path(p_preset));
+		if (!DirAccess::exists(globalized_template_path)) {
+			String current_version = VERSION_FULL_CONFIG;
+			String template_zip_path = EditorSettings::get_singleton()->get_templates_dir().plus_file(current_version).plus_file("android_source.zip");
+			ExportTemplateManager template_manager;
+			WARN_PRINT(globalized_template_path);
```
Comment (ellenhp) [inline platform/android/export/export_plugin.cpp]: Remove debug print.

File: servers/ramatak/ad_server.cpp [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+	} else if (p_location == BannerAdLocation::BANNER_LOCATION_TOP) {
+		location = "top";
+	} else {
+		ERR_FAIL_V_MSG(Variant(), "Ad location not handled");
+	}
+	Array ad_plugin_priorities = (Array)ProjectSettings::get_singleton()->get("ramatak/monetization/ad_plugin_priorities");
+	for (int i = 0; i < ad_plugin_priorities.size(); i++) {
+		String network_specific_id = MonetizationSettings::get_singleton()->get_ad_unit_network_id(p_ad_unit, ad_plugin_priorities[i]);
+		if (network_specific_id.length() > 0) {
+			WARN_PRINT_ONCE(ad_plugin_priorities[i]);
```
Comment (ellenhp) [inline servers/ramatak/ad_server.cpp]: Remove debug print.

File: core/os/dir_access.cpp [hunk 16]
```diff
@@ -327,13 +327,20 @@ Error DirAccess::copy(String p_from, String p_to, int p_chmod_flags) {
...
 	if (err == OK && p_chmod_flags != -1) {
-		fdst->close();
 		err = FileAccess::set_unix_permissions(p_to, p_chmod_flags);
 		// If running on a platform with no chmod support (i.e., Windows), don't fail
 		if (err == ERR_UNAVAILABLE) {
 			err = OK;
 		}
+	} else if (err == OK) {
+		WARN_PRINT("Setting old permissions");
```
Comment (ellenhp) [inline core/os/dir_access.cpp]: Remove debug print

File: editor/export_template_manager.cpp [hunk 17]
```diff
@@ -753,6 +753,108 @@ Error ExportTemplateManager::install_android_template_from_file(const String &p_
 	return OK;
 }
 
+Error ExportTemplateManager::install_android_template_from_file_to_path(const String &p_file, const String &p_path) {
+	// To support custom Android builds, we install the Java source code and buildsystem
+	// from android_source.zip to the project's user://android/VERSION folder.
+
+	WARN_PRINT(p_path);
```
Comment (ellenhp) [inline editor/export_template_manager.cpp]: Remove debug print.

File: editor/export_template_manager.cpp [hunk 18]
```diff
@@ -753,6 +753,108 @@ Error ExportTemplateManager::install_android_template_from_file(const String &p_
...
+
+	DirAccess::create(DirAccess::ACCESS_FILESYSTEM)->make_dir_recursive(p_path + "build");
```
Comment (ellenhp) [inline editor/export_template_manager.cpp]: Here and below: Replace concatenations with `plus_file` calls.

File: editor/export_template_manager.cpp [hunk 19]
```diff
@@ -753,6 +753,108 @@ Error ExportTemplateManager::install_android_template_from_file(const String &p_
 	return OK;
 }
 
+Error ExportTemplateManager::install_android_template_from_file_to_path(const String &p_file, const String &p_path) {
```
Comment (ellenhp) [inline editor/export_template_manager.cpp:756]: Consider consolidating logic with the function I copied this code from :)

File: doc/classes/AdController.xml [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="hide">
+			<return type="void" />
+			<description>
```
Comment (ellenhp) [inline doc/classes/AdController.xml:14]: Add documentation.

File: doc/classes/AdServer.xml [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<methods>
+		<method name="get_available_plugins" qualifiers="const">
+			<return type="Array" />
+			<description>
```
Comment (ellenhp) [inline doc/classes/AdServer.xml:14]: Add documentation.

File: doc/classes/BannerAdController.xml [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+	<brief_description>
+	</brief_description>
+	<description>
+	</description>
+	<tutorials>
+	</tutorials>
+	<methods>
+	</methods>
+	<members>
+		<member name="auto_show" type="bool" setter="set_auto_show" getter="get_auto_show" default="false">
```
Comment (ellenhp) [inline doc/classes/BannerAdController.xml:15]: Add documentation.

File: doc/classes/DummyAdPlugin.xml [hunk 23]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="DummyAdPlugin" inherits="AdPlugin" version="3.5" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+	</brief_description>
+	<description>
```
Comment (ellenhp) [inline doc/classes/DummyAdPlugin.xml:5]: Add documentation, or remove the dummy plugin altogether? We also could shift its role from no-op to logging maybe, but currently you can no-op by removing all of the plugins from the priority list.

File: doc/classes/ProjectSettings.xml [hunk 24]
```diff
@@ -1403,6 +1403,12 @@
 			[b]Note:[/b] Jitter fix is automatically disabled at runtime when [member physics/common/physics_interpolation] is enabled.
 			[b]Note:[/b] This property is only read when the project starts. To change the value at runtime, set [member Engine.physics_jitter_fix] instead.
 		</member>
+		<member name="ramatak/monetization/ad_plugin_config" type="Dictionary" setter="" getter="" default="{}">
```
Comment (ellenhp) [inline doc/classes/ProjectSettings.xml]: Here and below: Remove.

These should not show up in the main project settings menu.

File: editor/project_settings_editor.cpp [hunk 25]
```diff
@@ -2160,6 +2160,10 @@ ProjectSettingsEditor::ProjectSettingsEditor(EditorData *p_data) {
 	import_defaults_editor->set_name(TTR("Import Defaults"));
 	tab_container->add_child(import_defaults_editor);
 
+	ramatak_settings_editor = memnew(RamatakSettingsEditor);
+	ramatak_settings_editor->set_name(TTR("Ramatak"));
```
Comment (ellenhp) [inline editor/project_settings_editor.cpp:2164]: Consider changing the name/branding of the Ramatak settings tab.

File: core/os/dir_access.cpp (hunk 1 above)
Comment (ellenhp) [reply core/os/dir_access.cpp]: Done

File: core/os/dir_access.cpp (hunk 16 above)
Comment (ellenhp) [reply core/os/dir_access.cpp]: Done

File: doc/classes/ProjectSettings.xml (hunk 24 above)
Comment (ellenhp) [reply doc/classes/ProjectSettings.xml]: Done

File: editor/export_template_manager.cpp (hunk 19 above)
Comment (ellenhp) [reply editor/export_template_manager.cpp:756]: This ends up being unpleasant to do so I'll make an issue for it and resolve this comment.

File: editor/export_template_manager.cpp (hunk 17 above)
Comment (ellenhp) [reply editor/export_template_manager.cpp]: Done

File: editor/export_template_manager.cpp (hunk 18 above)
Comment (ellenhp) [reply editor/export_template_manager.cpp]: Done

File: editor/project_settings_editor.cpp (hunk 25 above)
Comment (ellenhp) [reply editor/project_settings_editor.cpp:2164]: This is fine for preview.

File: servers/ramatak/ad_server.cpp (hunk 7 above)
Comment (ellenhp) [reply servers/ramatak/ad_server.cpp]: Done

File: servers/ramatak/ad_server.cpp (hunk 15 above)
Comment (ellenhp) [reply servers/ramatak/ad_server.cpp]: Done

File: editor/ramatak/ramatak_settings_editor.cpp (hunk 12 above)
Comment (ellenhp) [reply editor/ramatak/ramatak_settings_editor.cpp]: Done

File: editor/ramatak/ramatak_settings_editor.cpp (hunk 13 above)
Comment (ellenhp) [reply editor/ramatak/ramatak_settings_editor.cpp]: Done

File: platform/android/export/export_plugin.cpp (hunk 11 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: Done

File: platform/android/export/export_plugin.cpp (hunk 2 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: Done

File: platform/android/export/export_plugin.cpp (hunk 3 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp:2323]: Ack.

File: platform/android/export/export_plugin.cpp (hunk 8 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: Done.

File: platform/android/export/export_plugin.cpp (hunk 14 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: Done.

File: platform/android/export/export_plugin.cpp (hunk 4 above)
Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: Ack

File: doc/classes/DummyAdPlugin.xml (hunk 23 above)
Comment (ellenhp) [reply doc/classes/DummyAdPlugin.xml:5]: I don't think we need to document this actually, since users should never interact with it.

File: servers/ramatak/ad_server.h (hunk 5 above)
Comment (ellenhp) [reply servers/ramatak/ad_server.h:19]: Ack

File: servers/ramatak/ad_server.h (hunk 6 above)
Comment (ellenhp) [reply servers/ramatak/ad_server.h:25]: Ack.

File: platform/android/export/export_plugin.cpp [hunk 26]
```diff
@@ -29,6 +29,7 @@
 /*************************************************************************/
 
 #include "export_plugin.h"
+#include "godot/editor/export_template_manager.h"
```
Comment (azagaya) [inline platform/android/export/export_plugin.cpp]: I had to change this include to `editor/export_template_manager.h` to be able to build

Comment (ellenhp) [reply platform/android/export/export_plugin.cpp]: I can't repro this build issue, which is confusing because the import is clearly wrong. Hate it when I can't figure out why something *does* work. Fixed.

File: platform/android/export/export_plugin.cpp [hunk 27]
```diff
@@ -222,8 +223,8 @@ static const LauncherIcon launcher_adaptive_icon_backgrounds[icon_densities_coun
 static const int EXPORT_FORMAT_APK = 0;
 static const int EXPORT_FORMAT_AAB = 1;
 
-static const char *APK_ASSETS_DIRECTORY = "res://android/build/assets";
-static const char *AAB_ASSETS_DIRECTORY = "res://android/build/assetPacks/installTime/src/main/assets";
+static const char *APK_ASSETS_DIRECTORY = "/build/assets";
```
Comment (hpvb) [inline platform/android/export/export_plugin.cpp:226]: Does this still work with a user-installed android template?

**Review by hpvb — APPROVED** [review]
